runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/graph/GreedilyFusedExecutableStage.java

```diff
@@ -0,0 +1,223 @@
+  private static final Logger LOG = LoggerFactory.getLogger(GreedilyFusedExecutableStage.class);
+
+  /**
+   * Returns an {@link ExecutableStage} where the initial {@link PTransformNode PTransform} is a
+   * Read which executes within the environment (with no input elements or {@link PCollectionNode
+   * PCollection}).
+   *
+   * <p>Once primitive Read transforms are replaced by an Impulse -> ParDo expansion, this
+   * construction should be removed.
+   */
+  public static ExecutableStage forRootTransform(
+      QueryablePipeline pipeline, PTransformNode rootTransform) {
+    checkArgument(
+        PTransformTranslation.READ_TRANSFORM_URN.equals(
+            rootTransform.getTransform().getSpec().getUrn()),
+        "Can only create an %s rooted at a %s node, got URN %s",
+        ExecutableStage.class.getSimpleName(),
+        PTransformTranslation.READ_TRANSFORM_URN,
+        rootTransform.getTransform().getSpec().getUrn());
+    return new GreedilyFusedExecutableStage(pipeline, null, Collections.singleton(rootTransform));
+  }
+
+  /**
+   * Returns an {@link ExecutableStage} where the initial {@link PTransformNode PTransform} is a
+   * Remote gRPC Port Read, reading elements from the materialized {@link PCollectionNode
+   * PCollection}.
+   *
+   * @param initialNodes the initial set of sibling transforms to fuse into this node. All of the
+   *     transforms must consume the {@code inputPCollection} on a per-element basis, and must all
+   *     be mutually compatible.
+   */
+  public static ExecutableStage forGrpcPortRead(
+      QueryablePipeline pipeline,
+      PCollectionNode inputPCollection,
+      Set<PTransformNode> initialNodes) {
+    return new GreedilyFusedExecutableStage(pipeline, inputPCollection, initialNodes);
+  }
+
+  private final QueryablePipeline pipeline;
+  private final Environment environment;
+
+  @Nullable private final PCollectionNode maybeInputPCollection;
+
+  private final Set<PCollectionNode> fusedCollections;
+  private final Set<PTransformNode> fusedTransforms;
+  private final Set<PCollectionNode> materializedPCollections;
+
+  private final Queue<PCollectionNode> fusionCandidates;
+
+  private GreedilyFusedExecutableStage(
+      QueryablePipeline pipeline,
+      @Nullable PCollectionNode inputPCollection,
+      Set<PTransformNode> initialNodes) {
+    checkArgument(
+        !initialNodes.isEmpty(),
+        "%s must contain at least one %s.",
+        GreedilyFusedExecutableStage.class.getSimpleName(),
+        PTransformNode.class.getSimpleName());
+    // Choose the environment from an arbitrary node. The initial nodes may not be empty for this
+    // subgraph to make any sense, there has to be at least one processor node
+    // (otherwise the stage is gRPC Read -> gRPC Write, which doesn't do anything).
+    this.environment =
+        pipeline
+            .getEnvironment(initialNodes.iterator().next())
```

Comment:
If the first node is a flatten, it won't have an environment since it doesn't need one.

You can flatten a single pcollection. (aka identity fn)

Reply:
Impulse + Flatten and Flatten with zero inputs are `valid` graphs though.

They usually come about because people use for loops in their pipeline construction.

Reply:
We can get to the edge cases when needed, most useful pipelines will never contain these constructs.

Reply:
My claim isn't that they're not valid within the overall pipeline, but they're not valid as the root transforms of an `ExecutableStage` (or at least not this one) - the pipeline should perform the `Flatten` (or other associated things) until a user stage begins.

Reply:
Makes sense that the Runner would be responsible for it and thus the PR #4647 will have to handle it.